### DISCUSSION TITLE
Can Legendary be used to install and update Unreal Engine on Windows?

### DISCUSSION THREAD

**Discussion by LRSFC-DanJ** [legendary-gl/legendary#241] [Q&A]
Epic's idea of academic deployment of Unreal Engine on multiple classroom machines is "use the Launcher to install it on one machine, then mirror that machine to the rest of your machines. Or, you can signup for our GitHub and compile it yourself". This doesn't work well for us as we haven't used mirroring/imaging as a process for deployment for many years now, and we don't have the resources or time to compile the software ourselves every time there is an update.

As far as I can tell Epic's Launcher doesn't provide any way to automate or script installations, but it looks like Legendary does.

So my question is, can Legendary be used to install/update Unreal Engine? On Windows specifically would be our interest.

**Comment by derrod** [comment]
Yes, by default UE is not included in the list, but the `--inckude-ue` flag will show you available versions that you can install. Be aware though that legendary does not currently support selective downloads for UE, so it would download everything which is larger than what you'd get with the epic launcher.

This can be worked around via the config and checking the install tags for the parts you need after doing an install via the Epic Launcher first. I believe updating should work fine with that as well (within one release, e.g. 4.26).

However since different engine releases (e.g. 4.25 and 4.26) are different products on the platform, updating from one to another is not possible and you would have to re-download the new release from scratch.

Better UE and also UE Marketplace support is on my todo list for the future, if you have any specific things that you're missing please leave some feedback.

**Reply by LRSFC-DanJ** [reply]
> This can be worked around via the config and checking the install tags for the parts you need after doing an install via the Epic Launcher first. I believe updating should work fine with that as well (within one release, e.g. 4.26).

Looking at the instructions, I'm guessing I need to do a `list-installed` after `auth`, to find out what the install tags are for the Epic Launcher version? And I can then use the `--install-tag` option of the `install` command to install only those items?

**Reply by derrod** [reply]
You'd have to do a `list-games --include-ue` to find the "app name" for the UE version you want. You can then use that with `list-files` to get the available install tags. However I would recommend to go on a machine that has the Epic Launcher installed and check e.g. the `ue426.sdmeta` file in `C:\ProgramData\Epic\EpicGamesLauncher\Data\EMS\EpicGamesLauncher`, it contains the descriptions for what each install tag actually is for.

Having just checked this, since "installing" is still disabled when install tag is manually used, patching wouldn't work. Instead any new update would have to be re-downloaded in its entireity. That is something I plan to change in the near future though.

**Reply by LRSFC-DanJ** [reply]
Now that I'm starting to look at how the process will work in more detail, it seems that the only methods for non-interactive (i.e. scripted) authentication are importing from the Epic Games Launcher or via an "exchange code", how does one get this code? Ideally I would not want to have to log on to a classroom full of 25 computers every time I need to install a new version of Unreal Engine.

EDIT: Wait, I see a `--download-only` option, does that mean I could download the installation files on a single central computer and then use some other method to install them on the target classroom computers? That would solve the authentication issue.

Also, is it possible to have multiple `--install-tag`s per `install` command, or do I need to do a separate `install` for each `--install-tag`?

**Reply by derrod** [reply]
Unfortunately epic's login system doesn't really allow for easy automation. You can get an exchange code via their API once you're logged in (those codes are what's given to game via launch parameters to login). Depending on how much work you want/can do you could potentially script something to generate exchange codes from a "master" system for each system to use to login when required. Also as long as you do anything that requires online authentication (even just running `legendary status`) every week or so the login session should be extended indefinitely.

Yes `--install-tag` is a parameter that can be specified multiple times. 

Download only mode is technically already enabled once you specify an install tag, so yeah, you could just download the UE data once and then copy it to all systems.

**Reply by LRSFC-DanJ** [reply]
Once I've got the files downloaded via `legendary install UE_4.26 --include-ue` with the various `--install-tag`s, how do I then install the downloaded files onto target systems? I don't see an .msi or a setup.exe or anything like that...

**Reply by derrod** [reply]
There's no installer, except maybe the one for the prerequisites. I don't actually have an UE stuff installed but I guess normally the Editor executable is just launched via the Epic launcher.

**Reply by LRSFC-DanJ** [reply]
Actually no - on Windows it creates a shortcut on the desktop that goes directly to the UE Editor.

However I've run into a snag. I noticed that what I got using the install tags is far smaller than what the Epic Games Launcher installed, so I had a look in the manifest that Legendary grabbed, and I found that some files, including the editor binary itself, don't have an install tag:

```
		{
			"Filename": "Engine/Binaries/Win64/UE4Editor.exe",
			"FileHash": "180247055083028161092232238111102122130169142116217124127255",
			"FileChunkParts": [
				{
					"Guid": "031EB66B46EEB79E2857138FD0D09065",
					"Offset": "183172009000",
					"Size": "073083006000"
				},
				{
					"Guid": "32FC196C4BDCECD1E4C49FB2970C7D0E",
					"Offset": "000000000000",
					"Size": "151091001000"
				}
			]
		}
```

So does that mean I have to download the whole thing, without using `--install-tag`?

**Reply by derrod** [reply]
Items without an install tags should work by also specifying the empty install tag via `--install-tag ""`

**Reply by LRSFC-DanJ** [reply]
Excellent. It looks like the following set of `--install-tag`s produces a set of files and folders that seems identical to the one produced by the Epic Games Launcher:

```
starter_content
templates
engine_source
platform_Android
platform_IOS
platform_Linux
""
```

I can then store this somewhere central and just use a script to robocopy it out to my target classroom machines.

**Reply by codecat** [reply]
They released Unreal Engine 5 Early Access. Is this installable via Legendary? I don't see it in the list when doing `list-games --include-ue`. The install URL on their site points to `com.epicgames.launcher://ue/ue5ea`.

**Reply by derrod** [reply]
You first have to add it to your account, which I'm not sure is currently possible via the website. So unfortunately you'll have to run the official launcher at least once to get it on your account.

**Reply by codecat** [reply]
Unfortunate, but that worked! It now shows up as:
```
UE5 Early Access (App name: UE_5.0EA | Version: 5.0.0-16433597+++UE5+Release-5.0-EarlyAccess-Windows)
```

**Comment by danj2k** [comment]
On a slight tangent, I've just noticed that it's possible to use the Windows version of Legendary to download stuff that's intended for the Mac platform by specifying `--platform Mac`, but how do you deal with the permissions issue? I can see the manifest contains flags like `bIsUnixExecutable`, are there any options for `list-files` that also output what permissions are required for each file? As this would also solve my problems in finding a way to deploy it to our Macs as well.

Also an update on the Windows situation, I was able to successfully deploy the Windows version of Unreal Engine 4.26.1 to a classroom of 24 machines using the set of files and folders Legendary downloaded.

**Reply by derrod** [reply]
The json output to `list-files` contains the `flags` property, if a file is supposed to be an executable the flag will be >= 4 (or `is_executable = flag & 0x4`). That is because the newer manifest format uses a combined flags field.

**Reply by LRSFC-DanJ** [reply]
Is there an explanation anywhere of what the various flags are? I can also see entries with `flags` values of 1, 2, 3, and 5 when I do `legendary.exe list-files UE_4.26 --platform Mac --json --install-tag "starter_content" --install-tag "templates" --install-tag "engine_source" --install-tag "platform_Android" --install-tag "platform_IOS" --install-tag "platform_Linux" --install-tag "" > UE_4_26-Mac.json`.

EDIT: nevermind, I found it:

```
    @property
    def read_only(self):
        return self.flags & 0x1

    @property
    def compressed(self):
        return self.flags & 0x2

    @property
    def executable(self):
        return self.flags & 0x4
```

I guess I don't need to care about things being compressed? So read-only is probably the only other one I need to worry about. And I assume by extension that means that anything that's not specifically read-only should be read-write?

So to check I've got this right, here's what I'm thinking:

```
flags == 0 means chmod 666
flags == 1 means chmod 644
flags == 2 means chmod 666
flags == 3 means chmod 644
flags == 4 means chmod 777
flags == 5 means chmod 755
```

Does that seem correct?

**Reply by derrod** [reply]
Seems correct, I don't know how the Mac version of the EGS handles file permissions since I never looked into it.
